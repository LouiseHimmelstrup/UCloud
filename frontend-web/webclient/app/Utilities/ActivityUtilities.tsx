import { buildQueryString } from "./URIUtilities";

export const activityQuery = (offset: number | null, scrollSize: number) => {
    const params = { scrollSize };
    if (offset !== null) params["offset"] = offset;

    return buildQueryString("/activity/browse/user", params);
}

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}` 

export const getLatestActivity = () => undefined;