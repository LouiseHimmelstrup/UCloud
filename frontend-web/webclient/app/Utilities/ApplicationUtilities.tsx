import { removeTrailingSlash } from "UtilityFunctions";
import { ParameterTypes, WithAppFavorite, WithAppMetadata, ApplicationParameter } from "Applications";
import Cloud from "Authentication/lib";
import { Page } from "Types";
import { expandHomeFolder } from "./FileUtilities";
import { AddSnackOperation, SnackType } from "Snackbar/Snackbars";
import swal, { SweetAlertResult } from "sweetalert2";

export const hpcJobQueryPost = "/hpc/jobs";

export const hpcJobQuery = (id: string, stdoutLine: number, stderrLine: number, stdoutMaxLines: number = 1000, stderrMaxLines: number = 1000) =>
    `/hpc/jobs/follow/${encodeURIComponent(id)}?stdoutLineStart=${stdoutLine}&stdoutMaxLines=${stdoutMaxLines}&stderrLineStart=${stderrLine}&stderrMaxLines=${stderrMaxLines}`;

export const hpcJobsQuery = (itemsPerPage: number, page: number): string =>
    `/hpc/jobs/?itemsPerPage=${itemsPerPage}&page=${page}`;

export const hpcFavoriteApp = (name: string, version: string) => `/hpc/apps/favorites/${encodeURIComponent(name)}/${encodeURIComponent(version)}`;

export const hpcFavorites = (itemsPerPage: number, pageNumber: number) =>
    `/hpc/apps/favorites?itemsPerPage=${itemsPerPage}&page=${pageNumber}`;

export const hpcApplicationsQuery = (page: number, itemsPerPage: number) =>
    `/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`;

interface HPCApplicationsSearchQuery { query: string, page: number, itemsPerPage: number }
export const hpcApplicationsSearchQuery = ({ query, page, itemsPerPage }): string =>
    `/hpc/apps/search?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const hpcApplicationsTagSearchQuery = ({ query, page, itemsPerPage }: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/searchTags?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const cancelJobQuery = `hpc/jobs`;


export function cancelJobSwal({ jobId }: { jobId: string }): Promise<SweetAlertResult> {
    return swal({
        title: "Cancel job?",
        text: `Cancel job: ${jobId}?`,
        allowEscapeKey: true,
        allowOutsideClick: true,
        allowEnterKey: false,
        showConfirmButton: true,
        showCancelButton: true,
        cancelButtonText: "No",
        confirmButtonText: "Cancel job",
    })
}

interface FavoriteApplicationFromPage extends AddSnackOperation {
    name: string
    version: string
    page: Page<WithAppMetadata & WithAppFavorite>
    cloud: Cloud
}
/**
* Favorites an application. 
* @param {Application} Application the application to be favorited
* @param {Cloud} cloud The cloud instance for requests
*/
export const favoriteApplicationFromPage = async ({ name, version, page, cloud, addSnack }: FavoriteApplicationFromPage): Promise<Page<WithAppMetadata & WithAppFavorite>> => {
    const a = page.items.find(it => it.metadata.name === name && it.metadata.version === version)!;
    try {
        await cloud.post(hpcFavoriteApp(name, version));
        a.favorite = !a.favorite;
    } catch {
        addSnack({ message: `An error ocurred favoriting ${name}`, type: SnackType.Failure });
    }
    return page;
}

type StringMap = { [k: string]: string }
interface AllowedParameterKey { name: string, type: ParameterTypes }
interface ExtractParameters {
    parameters: StringMap
    allowedParameterKeys: AllowedParameterKey[]
    siteVersion: number
}

export const extractParameters = ({ parameters, allowedParameterKeys, siteVersion }: ExtractParameters): StringMap => {
    let extractedParameters = {};
    if (siteVersion === 1) {
        allowedParameterKeys.forEach(({ name, type }) => {
            if (parameters[name] !== undefined) {
                if (compareType(type, parameters[name])) {
                    extractedParameters[name] = parameters[name];
                }
            }
        });
    }
    return extractedParameters;
}

export const isFileOrDirectoryParam = ({ type }: { type: string }) => type === "input_file" || type === "input_directory";


const compareType = (type: ParameterTypes, parameter: string): boolean => {
    switch (type) {
        case ParameterTypes.Boolean:
            return parameter === "Yes" || parameter === "No" || parameter === "";
        case ParameterTypes.Integer:
            return parseInt(parameter) % 1 === 0;
        case ParameterTypes.FloatingPoint:
            return typeof parseFloat(parameter) === "number";
        case ParameterTypes.Text:
        case ParameterTypes.InputDirectory:
        case ParameterTypes.InputFile:
            return typeof parameter === "string";

    }
}

interface ExtractedParameters {
    [key: string]: string | number | boolean | { source: string, destination: string }
}

export type ParameterValues = Map<string, React.RefObject<HTMLInputElement | HTMLSelectElement>>;

interface ExtractParametersFromMap {
    map: ParameterValues
    appParameters: ApplicationParameter[]
    cloud: Cloud
}

export function extractParametersFromMap({ map, appParameters, cloud }: ExtractParametersFromMap): ExtractedParameters {
    const extracted: ExtractedParameters = {};
    map.forEach(({ current }, key) => {
        const parameter = appParameters.find(it => it.name === key);
        if (!current) return;
        if (!current.value || !current.checkValidity()) return;
        if (!parameter) return;
        switch (parameter.type) {
            case ParameterTypes.InputDirectory:
            case ParameterTypes.InputFile:
                const expandedValue = expandHomeFolder(current.value, cloud.homeFolder)
                extracted[key] = {
                    source: expandedValue,
                    destination: removeTrailingSlash(expandedValue).split("/").pop()!
                };
                return;
            case ParameterTypes.Boolean:
                switch (current.value) {
                    case "Yes":
                        extracted[key] = true;
                        return;
                    case "No":
                        extracted[key] = false;
                        return;
                    default:
                        return;
                }
            case ParameterTypes.Integer:
                extracted[key] = parseInt(current.value);
                return;
            case ParameterTypes.FloatingPoint:
                extracted[key] = parseFloat(current.value);
                return;
            case ParameterTypes.Text:
                extracted[key] = current.value;
                return;
        }
    })
    return extracted;
}