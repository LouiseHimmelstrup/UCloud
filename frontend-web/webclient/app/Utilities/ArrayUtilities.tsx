export function removeEntry<T>(arr: T[], index: number):T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1)); 
}


type PrimitiveDataTypes = string | number | boolean
/**
 * @param set 
 * @param entry 
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size === set.size;
}