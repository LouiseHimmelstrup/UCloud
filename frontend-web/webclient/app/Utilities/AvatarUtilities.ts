import SDUCloud from "Authentication/lib";

export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
export const findAvatarBulk = (members: string[], cloud: SDUCloud) =>
    cloud.post(findAvatarBulkQuery, { usernames: members });

const findAvatarBulkQuery = "/avatar/find/bulk";
