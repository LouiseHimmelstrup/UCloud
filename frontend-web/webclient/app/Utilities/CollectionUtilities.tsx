type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
