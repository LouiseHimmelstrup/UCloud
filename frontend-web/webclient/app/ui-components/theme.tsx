const createMediaQuery = (n: string | number) => `@media screen and (min-width:${n})`

const addAliases = (arr: any, aliases: any[]) =>
  aliases.forEach((key, i) =>
    Object.defineProperty(arr, key, {
      enumerable: false,
      get() {
        return this[i]
      }
    })
  )

// export const breakpoints = [32, 40, 48, 64, 80].map(n => n + 'em')
export const breakpoints = [512, 640, 768, 1024, 1280].map(n => n + 'px')
const aliases = ['xs', 'sm', 'md', 'lg', 'xl']
export const responsiveBP = { xs: 512-1, sm: 640-1, md: 768-1, lg: 1024-1, xl: 1280-1 } 

export const mediaQueries = breakpoints.map(createMediaQuery)
addAliases(breakpoints, aliases)
addAliases(mediaQueries, aliases)

export const space = [0, 4, 8, 16, 32, 64, 128]

export const fontFamily = `'IBM Plex Sans', sans-serif`

export const fontSizes = [10, 14, 16, 20, 24, 32, 40, 56, 72]

export const medium = 300
export const bold = 700
export const regular = 400

// styled-system's `fontWeight` function can hook into the `fontWeights` object
export const fontWeights = {
  medium,
  bold,
  regular
}

export const lineHeights = {
  standard: 1.5,
  display: 1.25
}

const letterSpacings = {
  normal: 'normal',
  caps: '0.025em'
}

export const textStyles = {
  display8: {
    fontSize: fontSizes[8] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display7: {
    fontSize: fontSizes[7] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display6: {
    fontSize: fontSizes[6] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display5: {
    fontSize: fontSizes[5] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display4: {
    fontSize: fontSizes[4] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display3: {
    fontSize: fontSizes[3] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display2: {
    fontSize: fontSizes[2] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display1: {
    fontSize: fontSizes[1] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display0: {
    fontSize: fontSizes[0] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display,
    letterSpacing: letterSpacings.caps,
    textTransform: 'uppercase'
  },
  body2: {
    fontSize: fontSizes[2] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  },
  body1: {
    fontSize: fontSizes[1] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  },
  body0: {
    fontSize: fontSizes[0] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  }
}


// color palette
const black = "#000";
const white = "#fff";
const textBlack = "#1e252e"
//// Gray
const lightGray = "#ebeff3";
const midGray = "#c9d3df";
const gray = "#8393A7";
const darkGray = "#53657d";
//// Blue
const lightBlue = "#f0f6ff";
const lightBlue2 = "#cdf";
const blue = "#0055d5"; 
// const blue = "#007bff"; 
const darkBlue = "#049";
//// Green
const lightGreen = "#00ff77";
const green = "#00C05A"; // secondary
const darkGreen = "#00823c";
//// Red
const lightRed = "#fcc";
const red = "#c00"; // secondary
const darkRed = "#800";
//// Orange
const orange = "#fa0"; // secondary
const darkOrange = "#a50";
//// Purple
const lightPurple = "#ecf";
const purple = "#70b"; // secondary

// Colors in the array come in 3 shades: light, medium , dark
// last color is for logo centers only
const appColors = [
  //["#0096ff", "#043eff"], // blue
  ["#F7D06A", "#E98C33", "#C46927"], // gold
  ["#EC6F8E", "#C75480", "#AA2457"], // salmon
  ["#B8D1E3", "#7C8DB3", "#5B698C"], // silver
  ["#83D8F9", "#3F80F6", "#2951BE"], // blue
  ["#AE83CF", "#9065D1", "#68449E"], // violet
  ["#E392CC", "#E2689D", "#B33B6D"], // pink
  ["#ECB08C", "#EA7B4B", "#BC4F33"], // bronze
  ["#90DCA1", "#69C97D", "#4D9161"], // green
  ["#F3B576", "#B77D50", "#7C4C3C"], // brown
  ["#D57AC5", "#E439C9", "#A1328F"], // purple
  ["#98E0F9", "#53A5F5", "#3E79C0"], // lightblue
  ["#DC6AA6", "#C62A5A", "#AA2457"], // red
  ["#C9D3DF", "#8393A7", "#53657D"], // gray colors from the theme
];

// Color Themes
// should use colors from the palette above
//// Light color theme
// const text = "#001833";
const text = textBlack;
const textHighlight = blue;
const headerText = lightGray;
const headerBg = blue;
const headerIconColor = lightGray
const headerIconColor2 = midGray;
// const borderGray = "#d1d6db";
const borderGray = midGray; //used for borders of cards, pagination, sidebar
//const paginationHoverColor = "#f7f7f7"
const paginationHoverColor = lightBlue;
const paginationDisabled = lightGray;
// Icons colors
const iconColor = darkGray;
const iconColor2 = gray;
const FtIconColor = lightGray;
const FtIconColor2 = gray;
const spinnerColor = blue;
// File table colors
const tableRowHighlight = lightBlue;


const colors = {
  black,
  white,
  lightGray,
  midGray,
  gray,
  darkGray,
  blue,
  lightBlue,
  lightBlue2,
  darkBlue,
  green,
  lightGreen,
  darkGreen,
  red,
  lightRed,
  darkRed,
  orange,
  darkOrange,
  purple,
  lightPurple,
  text,
  textHighlight,
  headerText,
  headerBg,
  headerIconColor,
  headerIconColor2,
  borderGray,
  paginationHoverColor,
  paginationDisabled,
  iconColor,
  iconColor2,
  FtIconColor,
  FtIconColor2,
  spinnerColor,
  tableRowHighlight,
}

export type ThemeColor = keyof typeof colors;


export { colors }

export const colorStyles = {
  whiteOnText: {
    color: colors.white,
    backgroundColor: colors.text
  },
  whiteOnGray: {
    color: colors.white,
    backgroundColor: colors.gray
  },
  textOnLightGray: {
    color: colors.text,
    backgroundColor: colors.lightGray
  },
  whiteOnBlue: {
    color: colors.white,
    backgroundColor: colors.blue
  },
  blueOnLightBlue: {
    color: colors.blue,
    backgroundColor: colors.lightBlue
  },
  whiteOnGreen: {
    color: colors.white,
    backgroundColor: colors.green
  },
  greenOnLightGreen: {
    color: colors.green,
    backgroundColor: colors.lightGreen
  },
  whiteOnRed: {
    color: colors.white,
    backgroundColor: colors.red
  },
  redOnLightRed: {
    color: colors.red,
    backgroundColor: colors.lightRed
  },
  textOnOrange: {
    color: colors.text,
    backgroundColor: colors.orange
  },
  whiteOnPurple: {
    color: colors.white,
    backgroundColor: colors.purple
  },
  purpleOnLightPurple: {
    color: colors.purple,
    backgroundColor: colors.lightPurple
  },
  textOnWhite: {
    color: colors.text,
    backgroundColor: colors.white
  },
  grayOnWhite: {
    color: colors.gray,
    backgroundColor: colors.white
  },
  blueOnWhite: {
    color: colors.blue,
    backgroundColor: colors.white
  },
  greenOnWhite: {
    color: colors.green,
    backgroundColor: colors.white
  },
  redOnWhite: {
    color: colors.red,
    backgroundColor: colors.white
  },
  purpleOnWhite: {
    color: colors.purple,
    backgroundColor: colors.white
  },
  whiteOnDarkOrange: {
    color: colors.white,
    backgroundColor: colors.darkOrange
  },
  // info: textOnLightGray
  info: {
    color: colors.text,
    backgroundColor: colors.lightGray
  },
  // success: whiteOnGreen
  success: {
    color: colors.white,
    backgroundColor: colors.green
  },
  //warning: textOnOrange
  warning: {
    color: colors.text,
    backgroundColor: colors.orange
  },
  // danger: whiteOnRed
  danger: {
    color: colors.white,
    backgroundColor: colors.red
  }
}

// styled-system's `borderRadius` function can hook into the `radii` object/array
export const radii = [0, 2, 6]
export const radius = '5px'

export const maxContainerWidth = '1280px'

// boxShadows: styled-systems hooks into shadows
export const shadows = [
  `0 0 2px 0 rgba(0,0,0,.08),0 1px 4px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 2px 8px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 4px 16px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)`
]
const BoxShadowsAliases = ['sm', 'md', 'lg', 'xl'];
addAliases(shadows, BoxShadowsAliases);

// animation duration
export const duration = {
  fast: `150ms`,
  normal: `300ms`,
  slow: `450ms`,
  slowest: `600ms`
}

// animation easing curves
const easeInOut = 'cubic-bezier(0.5, 0, 0.25, 1)'
const easeOut = 'cubic-bezier(0, 0, 0.25, 1)'
const easeIn = 'cubic-bezier(0.5, 0, 1, 1)'

const timingFunctions = {
  easeInOut,
  easeOut,
  easeIn
}

// animation delay
const transitionDelays = {
  small: `60ms`,
  medium: `160ms`,
  large: `260ms`,
  xLarge: `360ms`
}

const theme = {
  breakpoints,
  mediaQueries,
  space,
  fontFamily,
  fontSizes,
  fontWeights,
  lineHeights,
  letterSpacings,
  regular,
  bold,
  textStyles,
  colors,
  colorStyles,
  appColors,
  radii,
  radius,
  shadows,
  maxContainerWidth,
  duration,
  timingFunctions,
  transitionDelays
}

export type Theme = typeof theme

export default theme
