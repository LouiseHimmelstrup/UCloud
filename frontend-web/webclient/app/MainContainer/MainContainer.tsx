import * as React from "react";
import { Box, Absolute } from "ui-components";
import * as Heading from "ui-components/Heading";
import { LoadableContent } from "LoadableContent";
import Spinner from "LoadingIcon/LoadingIcon";
import styled from "styled-components";
import { ReduxObject, ResponsiveReduxObject } from "DefaultObjects"
import { connect } from 'react-redux'

export interface MainContainerStateProps {
    responsiveState?: ResponsiveReduxObject
}

export interface MainContainerProps extends MainContainerStateProps {
    sidebar?: React.ReactNode, sidebarSize?: number,
    main?: React.ReactNode, additional?: React.ReactNode,
    header?: React.ReactNode, headerSize?: number,
}

export const _MainContainer = ({ sidebar, main, additional, header, sidebarSize = 240, headerSize = 96, responsiveState }: MainContainerProps) => {
    const leftSidebarsize = responsiveState!.greaterThan.xl ? 190 : 68; //main webside sidebar H size
    const topMenuSize = 48; //main website top manu V size
    const pad = 14; //padding unit

    const mainYpad = header ? headerSize : pad;
    const mainXpad = sidebar ? sidebarSize : pad;


    return (
        <React.StrictMode>
            <Box ml={leftSidebarsize} pt={topMenuSize} pb={pad} pl={pad} pr="0">
                {header &&
                    <HeaderContainer
                        top={topMenuSize} left="0"
                        py={pad} pl={leftSidebarsize + pad} pr={pad}
                        width={1} height={headerSize}
                        bg="white">
                        {header}
                    </HeaderContainer>
                }
                {sidebar &&
                    <SidebarContainer
                        height="100%" pt={topMenuSize + mainYpad}
                        top="0" right="0"
                        px={pad}
                        width={sidebarSize}
                    >
                        {sidebar}
                    </SidebarContainer>
                }
                <Box pt={mainYpad} pr={mainXpad}>
                    {main}
                </Box>
                {additional}
            </Box>
        </React.StrictMode>
    );
}

const mapStateToProps = ({ responsive }: ReduxObject): MainContainerStateProps => ({
    responsiveState: responsive
});

export const MainContainer = connect<MainContainerStateProps>(mapStateToProps)(_MainContainer);

export interface LoadingMainContainerProps<T = any> extends MainContainerProps {
    loadable: LoadableContent<T>
    fallbackHeader?: JSX.Element
    fallbackSidebar?: JSX.Element
}

export function LoadingMainContainer(props: LoadingMainContainerProps): JSX.Element {
    if (!props.loadable.content) {
        const main = !!props.loadable.error ?
            <Heading.h2>{props.loadable.error.statusCode} - {props.loadable.error.errorMessage}</Heading.h2> :
            <Spinner size={24} />
        return <MainContainer
            header={props.fallbackHeader}
            sidebar={props.fallbackSidebar}
            main={main} />;
    } else {
        return <MainContainer {...props} />
    }
}


const HeaderContainer = styled(Absolute)`
    position: fixed;
    z-index: 50;
`;

const SidebarContainer = styled(Absolute)`
    overflow-y: auto;
    position: fixed;
    z-index: 45;
`;