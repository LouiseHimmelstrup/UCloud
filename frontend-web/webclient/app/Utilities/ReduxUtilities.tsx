import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
