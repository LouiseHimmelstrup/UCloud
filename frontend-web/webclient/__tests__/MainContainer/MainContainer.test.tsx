import * as React from "react";
import {MainContainer} from "MainContainer/MainContainer";
import {create} from "react-test-renderer";
import {Provider} from "react-redux";
import {responsive, configureStore} from "Utilities/ReduxUtilities";
import "jest-styled-components";
import theme from "ui-components/theme";
import {ThemeProvider} from "styled-components";

describe("Main container", () => {
    it("Only main container", () =>
        expect(create(
            <Provider store={configureStore({}, {responsive})}>
                <ThemeProvider theme={theme}>
                    <MainContainer main={<div />} />
                </ThemeProvider>
            </Provider>
        ).toJSON()).toMatchSnapshot())

    it("Main container with sidebar", () =>
        expect(create(
            <Provider store={configureStore({}, {responsive})}>
                <ThemeProvider theme={theme}>
                    <MainContainer main={<div />} sidebar={<div />} />
                </ThemeProvider>
            </Provider>
        ).toJSON()).toMatchSnapshot())

    it("Main container with header", () =>
        expect(create(
            <Provider store={configureStore({}, {responsive})}>
                <ThemeProvider theme={theme}>
                    <MainContainer main={<div />} header={<div />} />
                </ThemeProvider>
            </Provider>
        ).toJSON()).toMatchSnapshot());
});