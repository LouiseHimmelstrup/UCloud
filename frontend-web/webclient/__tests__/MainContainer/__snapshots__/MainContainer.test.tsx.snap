// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main container Main container with header 1`] = `
.c0 {
  cursor: auto;
  margin-left: 190px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  padding-top: 48px;
}

.c2 {
  cursor: auto;
  padding-right: 14px;
  padding-top: 96px;
}

.c1 {
  cursor: auto;
  padding-left: 204px;
  padding-right: 14px;
  padding-top: 14px;
  padding-bottom: 14px;
  width: 100%;
  height: 96px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
    height={96}
    width={1}
  >
    <div />
  </div>
  <div
    className="c2"
    cursor="auto"
  >
    <div />
  </div>
</div>
`;

exports[`Main container Main container with sidebar 1`] = `
.c0 {
  cursor: auto;
  margin-left: 190px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  padding-top: 48px;
}

.c2 {
  cursor: auto;
  padding-right: 240px;
  padding-top: 14px;
}

.c1 {
  cursor: auto;
  padding-top: 62px;
  padding-left: 14px;
  padding-right: 14px;
  width: 240px;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  overflow-y: auto;
  position: fixed;
  z-index: 45;
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
    height="100%"
    width={240}
  >
    <div />
  </div>
  <div
    className="c2"
    cursor="auto"
  >
    <div />
  </div>
</div>
`;

exports[`Main container Only main container 1`] = `
.c0 {
  cursor: auto;
  margin-left: 190px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  padding-top: 48px;
}

.c1 {
  cursor: auto;
  padding-right: 14px;
  padding-top: 14px;
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
  >
    <div />
  </div>
</div>
`;
